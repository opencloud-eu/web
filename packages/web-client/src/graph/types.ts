import type { AxiosInstance, ResponseType } from 'axios'
import type { Configuration } from './generated'

export interface GraphFactoryOptions {
  axiosClient: AxiosInstance
  config: Configuration
}

export interface GraphRequestOptions {
  headers?: Record<string, string>
  params?: Record<string, string>
  signal?: AbortSignal
  responseType?: ResponseType
}
