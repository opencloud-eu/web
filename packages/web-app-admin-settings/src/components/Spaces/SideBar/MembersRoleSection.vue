<template>
  <ul class="oc-list">
    <li
      v-for="(m, index) in members"
      :key="index"
      class="oc-flex oc-flex-middle oc-mb-s"
      data-testid="space-members-list"
    >
      <oc-avatar
        v-if="m.grantedTo.user"
        :user-name="m.grantedTo.user.displayName"
        :width="36"
        class="oc-mr-s"
      /><oc-avatar-item
        v-else
        :width="36"
        icon-size="medium"
        :icon="groupIcon"
        name="group"
        class="oc-mr-s"
      />
      {{ (m.grantedTo.user || m.grantedTo.group).displayName }}
    </li>
  </ul>
</template>
<script lang="ts">
import { computed, defineComponent, PropType } from 'vue'
import { ShareTypes, SpaceMember } from '@opencloud-eu/web-client'

export default defineComponent({
  name: 'MembersRoleSection',
  props: {
    members: {
      type: Array as PropType<SpaceMember[]>,
      required: true
    }
  },
  setup() {
    const groupIcon = computed(() => {
      return ShareTypes.group.icon
    })
    return { groupIcon }
  }
})
</script>
