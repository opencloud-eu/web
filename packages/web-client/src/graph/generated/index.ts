/* tslint:disable */
/* eslint-disable */
/**
 * Libre Graph API
 * Libre Graph is a free API for cloud collaboration inspired by the MS Graph API.
 *
 * The version of the OpenAPI document: v1.0.7
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


export * from "./api";
export * from "./configuration";

