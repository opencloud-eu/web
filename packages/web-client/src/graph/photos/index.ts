export * from './photo'
export * from './types'
