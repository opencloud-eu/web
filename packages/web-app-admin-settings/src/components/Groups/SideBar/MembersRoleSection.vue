<template>
  <ul class="oc-list">
    <li
      v-for="(member, index) in groupMembers"
      :key="index"
      class="oc-flex oc-flex-middle oc-mb-s"
      data-testid="group-members-list"
    >
      <oc-avatar :user-name="member.displayName" :width="36" class="oc-mr-s" />
      <span class="oc-text-truncate" :title="member.displayName">
        {{ member.displayName }}
      </span>
    </li>
  </ul>
</template>
<script lang="ts">
import { defineComponent, PropType } from 'vue'
import { User } from '@opencloud-eu/web-client/graph/generated'

export default defineComponent({
  name: 'MembersRoleSection',
  props: {
    groupMembers: {
      type: Array as PropType<User[]>,
      required: false,
      default: (): User[] => []
    }
  }
})
</script>
