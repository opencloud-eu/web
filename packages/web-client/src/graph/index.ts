import { AxiosInstance } from 'axios'
import { Configuration } from './generated'
import { type GraphUsers, UsersFactory } from './users'
import { type GraphGroups, GroupsFactory } from './groups'
import { ApplicationsFactory, GraphApplications } from './applications'
import { DrivesFactory, GraphDrives } from './drives'
import { DriveItemsFactory, GraphDriveItems } from './driveItems'
import { TagsFactory, GraphTags } from './tags'
import { ActivitiesFactory, GraphActivities } from './activities'
import { PermissionsFactory, GraphPermissions } from './permissions'
import { PhotoFactory, GraphPhotos } from './photos'

export interface Graph {
  activities: GraphActivities
  applications: GraphApplications
  tags: GraphTags
  drives: GraphDrives
  driveItems: GraphDriveItems
  users: GraphUsers
  groups: GraphGroups
  permissions: GraphPermissions
  photos: GraphPhotos
}

export const graph = (baseURI: string, axiosClient: AxiosInstance): Graph => {
  const url = new URL(baseURI)
  url.pathname = [...url.pathname.split('/'), 'graph'].filter(Boolean).join('/')
  const config = new Configuration({
    basePath: url.href
  })

  return <Graph>{
    activities: ActivitiesFactory({ axiosClient, config }),
    applications: ApplicationsFactory({ axiosClient, config }),
    tags: TagsFactory({ axiosClient, config }),
    drives: DrivesFactory({ axiosClient, config }),
    driveItems: DriveItemsFactory({ axiosClient, config }),
    users: UsersFactory({ axiosClient, config }),
    groups: GroupsFactory({ axiosClient, config }),
    permissions: PermissionsFactory({ axiosClient, config }),
    photos: PhotoFactory({ axiosClient, config })
  }
}
