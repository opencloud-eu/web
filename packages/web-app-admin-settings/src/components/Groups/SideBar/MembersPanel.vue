<template>
  <div class="oc-ml-s">
    <oc-text-input
      v-model="filterTerm"
      class="oc-text-truncate oc-mr-s oc-mt-m"
      :label="$gettext('Filter members')"
    />
    <div ref="membersListRef" data-testid="space-members">
      <div v-if="!filteredGroupMembers.length">
        <h3 class="oc-text-bold oc-text-medium" v-text="$gettext('No members found')" />
      </div>
      <div v-if="filteredGroupMembers.length" class="oc-mb-m" data-testid="group-members">
        <h3 class="oc-text-bold oc-text-medium" v-text="$gettext('Members')" />
        <members-role-section :group-members="filteredGroupMembers" />
      </div>
    </div>
  </div>
</template>
<script lang="ts">
import { computed, defineComponent, inject, ref, watch, unref, Ref } from 'vue'
import MembersRoleSection from '../../Groups/SideBar/MembersRoleSection.vue'
import Fuse from 'fuse.js'
import Mark from 'mark.js'
import { Group, User } from '@opencloud-eu/web-client/graph/generated'
import { defaultFuseOptions } from '@opencloud-eu/web-pkg'

export default defineComponent({
  name: 'GroupsMembersPanel',
  components: { MembersRoleSection },
  setup() {
    const group = inject<Ref<Group>>('group')
    const filterTerm = ref('')
    const markInstance = ref(null)
    const membersListRef = ref(null)

    const filterMembers = (collection: User[], term: string) => {
      if (!(term || '').trim()) {
        return collection
      }

      const searchEngine = new Fuse(collection, { ...defaultFuseOptions, keys: ['displayName'] })
      return searchEngine.search(term).map((r) => r.item)
    }

    const members = computed(() => {
      if (group) {
        return unref(group).members.sort((a, b) => a.displayName.localeCompare(b.displayName))
      }
      return []
    })

    const filteredGroupMembers = computed(() => {
      return filterMembers(unref(members), unref(filterTerm))
    })

    watch(filterTerm, () => {
      if (unref(membersListRef)) {
        markInstance.value = new Mark(unref(membersListRef))
        unref(markInstance).unmark()
        const searchTermRegex = unref(filterTerm)
        unref(markInstance).mark(searchTermRegex, {
          element: 'span',
          className: 'mark-highlight'
        })
      }
    })

    return {
      filterTerm,
      membersListRef,
      members,
      filteredGroupMembers
    }
  }
})
</script>
