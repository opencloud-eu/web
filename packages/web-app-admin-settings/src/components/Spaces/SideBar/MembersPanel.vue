<template>
  <div class="oc-ml-s">
    <oc-text-input
      v-model="filterTerm"
      class="oc-mr-s oc-mt-m"
      :label="$gettext('Filter members')"
    />
    <div ref="membersListRef" data-testid="space-members">
      <div v-if="!filteredPermissions.length">
        <h3 class="oc-text-bold oc-text-medium" v-text="$gettext('No members found')" />
      </div>
      <div v-for="(role, i) in availableRoles" :key="i">
        <div
          v-if="getPermissionsForRole(role).length"
          class="oc-mb-m"
          :data-testid="`space-members-role-${role.displayName}`"
        >
          <h3 class="oc-text-bold oc-text-medium" v-text="role.displayName" />
          <members-role-section :permissions="getPermissionsForRole(role)" />
        </div>
      </div>
      <div v-if="permissionsWithoutRole.length" class="space-members-custom">
        <h3 class="oc-text-bold oc-text-medium" v-text="$gettext('Custom role')" />
        <members-role-section :permissions="permissionsWithoutRole" />
      </div>
    </div>
  </div>
</template>
<script setup lang="ts">
import { computed, inject, ref, watch, unref } from 'vue'
import { ShareRole, SpaceResource } from '@opencloud-eu/web-client'
import MembersRoleSection from './MembersRoleSection.vue'
import Fuse from 'fuse.js'
import Mark from 'mark.js'
import { defaultFuseOptions, useSharesStore } from '@opencloud-eu/web-pkg'
import { Permission } from '@opencloud-eu/web-client/graph/generated'

const sharesStore = useSharesStore()

const resource = inject<SpaceResource>('resource')
const filterTerm = ref('')
const markInstance = ref(null)
const membersListRef = ref(null)

const filterMembers = (collection: Permission[], term: string) => {
  if (!(term || '').trim()) {
    return collection
  }

  const searchEngine = new Fuse(collection, {
    ...defaultFuseOptions,
    keys: ['grantedToV2.user.displayName', 'grantedToV2.group.displayName']
  })
  return searchEngine.search(term).map((r) => r.item)
}

const permissions = computed(() => {
  return Object.values(unref(resource).root.permissions)
})

const filteredPermissions = computed(() => {
  return filterMembers(unref(permissions), unref(filterTerm))
})

const availableRoles = computed<ShareRole[]>(() => {
  const permissionsWithRole = unref(permissions).filter((p) => !!p.roles.length)
  const roleIds = [...new Set(permissionsWithRole.map((p) => p.roles).flat())]
  return roleIds
    .map((r) => sharesStore.graphRoles[r])
    .filter(Boolean)
    .sort((a, b) => {
      // sort roles by amount of permissions (most likely translates to manager > editor > viewer)
      const permissionsA = a.rolePermissions.flatMap((r) => r.allowedResourceActions)
      const permissionsB = b.rolePermissions.flatMap((r) => r.allowedResourceActions)
      return permissionsB.length - permissionsA.length
    })
})

const permissionsWithoutRole = computed(() => {
  return unref(filteredPermissions).filter(({ roles }) => !roles.length)
})

const getPermissionsForRole = (role: ShareRole) => {
  return unref(filteredPermissions).filter(({ roles }) => roles.includes(role.id))
}

watch(filterTerm, () => {
  if (unref(membersListRef)) {
    markInstance.value = new Mark(unref(membersListRef))
    unref(markInstance).unmark()
    unref(markInstance).mark(unref(filterTerm), {
      element: 'span',
      className: 'mark-highlight'
    })
  }
})
</script>
