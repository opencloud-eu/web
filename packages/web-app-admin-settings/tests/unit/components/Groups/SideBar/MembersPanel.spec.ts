import MembersPanel from '../../../../../src/components/Groups/SideBar/MembersPanel.vue'
import { defaultPlugins, shallowMount } from '@opencloud-eu/web-test-helpers'
import { mock } from 'vitest-mock-extended'
import { Group } from '@opencloud-eu/web-client/graph/generated'
import MembersRoleSection from '../../../../../src/components/Groups/SideBar/MembersRoleSection.vue'

const groupMock = mock<Group>({
  id: '1',
  groupTypes: [],
  members: [{ displayName: 'Albert Einstein' }]
})
const selectors = {
  membersRolePanelStub: 'members-role-section-stub'
}

describe('MembersPanel', () => {
  it('should render all members accordingly to their role assignments', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should filter members accordingly to the entered search term', async () => {
    const { wrapper } = getWrapper()
    wrapper.vm.filterTerm = 'ein'
    await wrapper.vm.$nextTick
    expect(wrapper.findAll(selectors.membersRolePanelStub).length).toBe(1)
    expect(
      wrapper.findComponent<typeof MembersRoleSection>(selectors.membersRolePanelStub).props()
        .groupMembers[0].displayName
    ).toEqual('Albert Einstein')
  })
  it('should display an empty result if no matching members found', async () => {
    const { wrapper } = getWrapper()
    wrapper.vm.filterTerm = 'no-match'
    await wrapper.vm.$nextTick
    expect(wrapper.findAll(selectors.membersRolePanelStub).length).toBe(0)
    expect(wrapper.html()).toMatchSnapshot()
  })
})

function getWrapper({ group = groupMock } = {}) {
  return {
    wrapper: shallowMount(MembersPanel, {
      global: {
        plugins: [...defaultPlugins()],
        provide: { group: group }
      }
    })
  }
}
