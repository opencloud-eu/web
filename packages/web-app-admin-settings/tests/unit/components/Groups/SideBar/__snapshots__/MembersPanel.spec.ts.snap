// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MembersPanel > should display an empty result if no matching members found 1`] = `
"<div class="oc-ml-s">
  <oc-text-input-stub label="Filter members" id="oc-textinput-3" type="text" modelvalue="no-match" clearbuttonenabled="false" clearbuttonaccessiblelabel="" disabled="false" fixmessageline="false" readonly="false" passwordpolicy="[object Object]" class="oc-mr-s oc-mt-m"></oc-text-input-stub>
  <div data-testid="space-members">
    <div>
      <h3 class="oc-text-bold oc-text-medium">No members found</h3>
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`MembersPanel > should render all members accordingly to their role assignments 1`] = `
"<div class="oc-ml-s">
  <oc-text-input-stub label="Filter members" id="oc-textinput-1" type="text" modelvalue="" clearbuttonenabled="false" clearbuttonaccessiblelabel="" disabled="false" fixmessageline="false" readonly="false" passwordpolicy="[object Object]" class="oc-mr-s oc-mt-m"></oc-text-input-stub>
  <div data-testid="space-members">
    <!--v-if-->
    <div class="oc-mb-m" data-testid="group-members">
      <h3 class="oc-text-bold oc-text-medium">Members</h3>
      <members-role-section-stub groupmembers="undefined"></members-role-section-stub>
    </div>
  </div>
</div>"
`;
